File: java/jswap/swap/SwapMote.java
/**
 * SwapMote.java
 * 
 * Copyright (c) 2011 Daniel Berenguer <dberenguer@usapiens.com>
 *  
 * This file is part of the panStamp project.
 *  
 * panStamp  is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * any later version.
 * 
 * panLoader is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with panLoader; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 
 * USA
 * 
 * Author: Daniel Berenguer
 * Creation date: 04/01/2011
 */
package swap;

import device.Settings;
import xmltools.XmlDevice;
import ccexception.CcException;
import device.SwapGateway;
import xmltools.XmlException;

import java.io.File;

/**
 * Class: SwapMote
 * 
 * Description:
 * 
 * Basic SWAP mote class
 */
public class SwapMote
{
  /**
   * Serial gateway
   */
  private static SwapGateway gateway;

  /**
   * Manufacturer id
   */
  private long manufactId = 0;

  /**
   * Product identifier
   */
  private long productId = 0;

  /**
   * Device address
   */
  private int address;

  /**
   * Cyclic nonce
   */
  private int nonce = 0;

  /**
   * System state
   */
  private int state = SwapDefs.SYSTATE_RUNNING;

  /**
   * Mote definition
   */
  private XmlDevice definition;

  /**
   * SwapMote
   * 
   * Class constructor
   * 
   * 'productCode'	Product code (Manufact ID + Product ID)
   * 'address'	Device address
   */
  public SwapMote(int[] productCode, int address) throws XmlException
  {
    // Read general settings from XML file
    Settings.read();
    
    int i;
    for(i=0 ; i<4 ; i++)
      manufactId |= productCode[i] << 8*(3-i);
    for(i=0 ; i<4 ; i++)
      productId |= productCode[i+4] << 8*(3-i);

    this.address = address;
    this.definition = new XmlDevice(Settings.getDeviceDir() + File.separator + Long.toHexString(manufactId) + File.separator + Long.toHexString(productId) + ".xml");
  }

  /**
   * getPwrDownMode
   * 
   * Return power-down flag
   */
  public boolean getPwrDownMode()
  {
    return definition.getPwrDownMode();
  }

  /**
   * getManufacturer
   * 
   * Return manufacturer string
   */
  public String getManufacturer() 
  {
    return definition.getManufacturer();
  }

  /**
   * getProduct
   * 
   * Return product string
   */
  public String getProduct() 
  {
    return definition.getProduct();
  }

  /**
   * getManufactId
   *
   * Get manufacturer id
   */
  public final long getManufactId()
  {
    return manufactId;
  }

  /**
   * getProductId
   *
   * Get product id
   */
  public final long getProductId()
  {
    return productId;
  }

  /**
   * getAddress
   *
   * Get device address
   */
  public final int getAddress()
  {
    return address;
  }

  /**
   * setAddress
   *
   * Set device address
   *
   * 'value'	Device address
   */
  public void setAddress(int value)
  {
    address = value;
  }

  /**
   * getNonce
   *
   * Get current cyclic nonce
   */
  public final int getNonce()
  {
    return nonce;
  }

  /**
   * setNonce
   *
   * Set cyclic nonce value
   */
  public void setNonce(int value)
  {
    nonce = value;
  }
  
  /**
   * incrementNonce
   *
   * Increment cyclic nmonce by one
   */
  public void incrementNonce()
  {
    nonce++;
  }

  /**
   * sendFreqChannelWack
   *
   * Send frequency channel value and wait for response from the mote
   *
   * 'freqChannel'	New frequency channel
   *
   * Return true if the mote confirmed (ACK'ed) the new frequency channel
   */
  public boolean sendFreqChannelWack(int freqChannel) throws ccexception.CcException
  {
    SwapValue val = new SwapValue(freqChannel, 1);
    return gateway.setMoteRegister(this, SwapDefs.ID_FREQ_CHANNEL, val);
  }

  /**
   * sendNetworkIdWack
   *
   * Send network id value and wait for response from the mote
   *
   * 'netId'	New network id
   *
   * Return true if the mote confirmed (ACK'ed) the new network id
   */
  public boolean sendNetworkIdWack(int netId) throws CcException
  {
    SwapValue val = new SwapValue(netId, 2);
    return gateway.setMoteRegister(this, SwapDefs.ID_NETWORK_ID, val);
  }

  /**
   * sendSecurityWack
   *
   * Send security option value and wait for response from the mote
   *
   * 'secu'	Security option
   *
   * Return true if the mote confirmed (ACK'ed) the new security option
   */
  public boolean sendSecurityWack(int secu) throws CcException
  {
    SwapValue val = new SwapValue(secu, 1);
    return gateway.setMoteRegister(this, SwapDefs.ID_SECU_OPTION, val);
  }

  /**
   * sendAddressWack
   *
   * Send address value and wait for response from the mote
   *
   * 'addr'	New device address
   *
   * Return true if the mote confirmed (ACK'ed) the new device address
   */
  public boolean sendAddressWack(int addr) throws CcException
  {
    SwapValue val = new SwapValue(addr, 1);
    return gateway.setMoteRegister(this, SwapDefs.ID_DEVICE_ADDR, val);
  }

  /**
   * restart
   *
   * Restart wireless mote
   *
   * Return true if the mote confirmed (ACK'ed) the command
   */
  public boolean restart() throws CcException
  {
    SwapValue val = new SwapValue(SwapDefs.SYSTATE_RESTART, 1);
    return gateway.setMoteRegister(this, SwapDefs.ID_SYSTEM_STATE, val);
  }
  
  /**
   * cmdRegister
   *
   * Send command to register
   *
   * 'id'   Register ID
   * 'val'	New endpoint value
   *
   * Return expected response to be received from the targeted endpoint
   */
  public SwapInfoPacket cmdRegister(int id, SwapValue val) throws CcException
  {
    SwapInfoPacket infPacket = new SwapInfoPacket(this.getAddress(), id, val);
    SwapCommandPacket cmdPacket = new SwapCommandPacket(this.nonce, this.getAddress(), id, val);
    cmdPacket.send();

    return infPacket;
  }

  /**
   * qryRegister
   *
   * Send query to register
   *
   * 'id'	Register ID
   */
  public void qryRegister(int id) throws CcException
  {
    SwapQueryPacket qryPacket = new SwapQueryPacket(this.getAddress(), id);
    qryPacket.send();
  }
  
  /**
   * getState
   *
   * Return system state
   */
  public final int getState()
  {
    return state;
  }

  /**
   * setState
   *
   * Set new system state
   */
  public void setState(int value)
  {
    state = value;
  }

  /**
   * getGateway
   *
   * Return serial gateway
   */
  public final static SwapGateway getGateway()
  {
    return gateway;
  }

  /**
   * setGateway
   *
   * Set serial gateway attribute
   */
  public static void setGateway(SwapGateway value)
  {
    gateway = value;
  }
}

File: python/pySwap/swap/SwapMote.py
#########################################################################
#
# SwapMote
#
# Copyright (c) 2011 Daniel Berenguer <dberenguer@usapiens.com>
# 
# This file is part of the panStamp project.
# 
# panStamp  is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# panStamp is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of 
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with panLoader; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 
# USA
#
# Author: Daniel Berenguer
# Creation date: 20-Aug-2011
#
#########################################################################
__author__="Daniel Berenguer"
__date__  ="$Aug 20, 2011 10:36:00 AM$"
#########################################################################

from swap.SwapInfoPacket import SwapInfoPacket
from swap.SwapCommandPacket import SwapCommandPacket
from swap.SwapQueryPacket import SwapQueryPacket
from swap.SwapDefs import SwapRegId, SwapState
from swap.SwapValue import SwapValue
from swapexception.SwapException import SwapException
from xmltools.XmlDevice import XmlDevice

import time


class SwapMote(object):
    """ SWAP device class"""
    def cmdRegister(self, regId, value):
        """ Send command to register and return expected response (SWAP info)"""
        # Expected response from mote
        infPacket = SwapInfoPacket(self.address, regId, value)
        # Command to be sent to the mote
        cmdPacket = SwapCommandPacket(self.address, regId, value, self.nonce)
        # Send command
        cmdPacket.send(self.server.modem)
        # Return expected response
        return infPacket;


    def qryRegister(self, regId):
        """ Send query to register"""
        # Query packet to be sent
        qryPacket = SwapQueryPacket(self.address, regId)
        # Send query
        qryPacket.send(self.server.modem)


    def infRegister(self, regId):
        """ Send SWAP info packet about the current value of the register passed as argument"""
        # Info packet to be sent
        infPacket = SwapInfoPacket(self.address, regId)
        # Send SWAP info packet
        infPacket.send(self.server.modem)


    def cmdRegisterWack(self, regId, value):
        """ Send SWAP command to remote register and wait for confirmation
        Return True if ACK received from mote """
        return self.server.setMoteRegister(self, regId, value)


    def setAddress(self, address):
        """ Set mote address. Return true if ACK received from mote"""
        val = SwapValue(address, length=1)
        return self.cmdRegisterWack(SwapRegId.ID_DEVICE_ADDR, val)


    def setNetworkId(self, netId):
        """ Set mote's network id. Return true if ACK received from mote"""
        val = SwapValue(netId, length=2)
        return self.cmdRegisterWack(SwapRegId.ID_NETWORK_ID, val)


    def setFreqChannel(self, channel):
        """ Set mote's frequency channel. Return true if ACK received from mote"""
        val = SwapValue(channel, length=1)
        return self.cmdRegisterWack(SwapRegId.ID_FREQ_CHANNEL, val)


    def setSecurity(self, secu):
        """ Set mote's security option. Return true if ACK received from mote"""
        val = SwapValue(secu, length=1)
        return self.cmdRegisterWack(SwapRegId.ID_SECU_OPTION, val)

    
    def restart(self):
        """ Ask mote to restart """
        val = SwapValue(SwapState.RESTART, length=1)
        return self.cmdRegisterWack(SwapRegId.ID_SYSTEM_STATE, val)


    def leaveSync(self):
        """ Ask mote to leave SYNC mode """
        val = SwapValue(SwapState.STOP, length=1)
        return self.cmdRegisterWack(SwapRegId.ID_SYSTEM_STATE, val)

    
    def updateTimeStamp(self):
        """
        Update time stamp
        """
        self.timeStamp = time.time()
    
    
    def __init__(self, server=None, productCode=None, address=0xFF):
        if server is None:
            raise SwapException("SwapMote constructor needs a valid SwapServer object")
        # Swap server object
        self.server = server
        # Product ID
        self.productId = 0
        # Manufacturer ID
        self.manufacturerId = 0
        # Definition settings
        self.config = None

        # Get manufacturer and product id from product code
        if productCode is not None:
            for i in range(4):
                self.manufacturerId = self.manufacturerId | (productCode[i] << 8 * (3-i))
                self.productId = self.productId | (productCode[i + 4] << 8 * (3-i))

        # Definition file
        # Definition settings
        self.definition = XmlDevice(self);

        # Device address
        self.address = address
        # Current mote's security nonce
        self.nonce = 0
        # State of the mote
        self.state = SwapState.RUNNING
        # List of regular registers provided by this mote
        self.lstRegRegs = None
        # List of config registers provided by this mote
        self.lstCfgRegs = None
        if self.definition is not None:
            # List of regular registers
            self.lstRegRegs = self.definition.getRegList()
            # List of config registers
            self.lstCfgRegs = self.definition.getRegList(config=True)
        # Initial time stamp
        self.timeStamp = None

